DRILL-7115: Improve Hive schema show tables performance

ihuzenko opened the pull request:
1. To make SHOW TABLES for Hive schema work much faster, additional Drill
   feature of showing only accesible tables when Storage-Based authorization
   is enabled was sacrificed. Now the behaviour matches to Hive/Beeline, all
   tables will be shown despite of accessibility. For details about previous
   show tables results, check description of DRILL-540.
2. In HiveDatabaseSchema implemented faster getTableNamesAndTypes() method
   and removed bulk related code.
3. Deprecated bulk related options and removed bulk code from AbstractSchema,
   DrillHiveMetastoreClient.
4. For 8000 Hive tables query returned in 1.8 seconds, for combination of
   4000 tables and 8000 views query returned in 2.3 seconds. Note, that
   after first query table names will be cached and next queries will perform
   in less than 1 sec.
5. Refactored WorkspaceSchemaFactory's getTableNamesAndTypes()
   method to reuse existing getViews() method.
6. DrillHiveMetastoreClient was refactored. Classes were unnested and enclosed
   within client package with restricted visibility. Also was updated cache
   values type to avoid unnecessarry List to Set back and forth conversions.
   Client creation methods moved to separate class. So the new package
   exposes only factory and client class.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java:
```diff
@@ -275,29 +274,14 @@ public void dropTable(String tableName) {
   public List<Pair<String, ? extends Table>> getTablesByNames(final List<String> tableNames) {
-    final List<Pair<String, ? extends Table>> tables = Lists.newArrayList();
+    final List<Pair<String, ? extends Table>> tables = new ArrayList<>(tableNames.size());
     for (String tableName : tableNames) {
```
```
    return tableNames.stream()
        // Schema may return NULL for table if the query user doesn't have permissions to load the table. Ignore such
        // tables as INFO SCHEMA is about showing tables which the use has access to query.
        .map(tableName -> Pair.of(tableName, getTable(tableName)))
        .filter(pair -> Objects.nonNull(pair.getRight()))
        .collect(Collectors.toList());
```

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java:
```diff
@@ -275,29 +274,14 @@ public void dropTable(String tableName) {
    * Get the collection of {@link Table} tables specified in the tableNames.
    *
-   * @param  tableNames the requested tables, specified by the table names
+   * @param  tableNames the requested tables, specified by the table namesbulkSize
```
typo?

ihuzenko commented:
Hi @vdiravka , according to the [PR discussion](https://github.com/apache/drill/pull/461#discussion_r58753354) , show tables authorization was left in inconsistent state (all tables shown for SQL Standard mode, only accessible shown for Storage Based mode). I have spent some time investigating the issues and even found how to fix listing for SQL Standard mode. But performance acceptable solution for Storage Based mode is still unknown. I think that listing of only accessible tables wasn't Drill users requirement, and in case of such Jira appearance it should be implemented as finalized feature with acceptable performance and consistency for both authorization modes. Until Drill users require the feature, it'll be much better to return all tables as fast as possible.

ihuzenko commented:
@vdiravka could you please review again from scratch ? Because a lot of things were changed since last review.

vdiravka commented on contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/client/TableName.java:
```diff
@@ -0,0 +1,73 @@
+final class TableName {
+
+  private final String dbName;
+
```
```suggestion
```

vdiravka commented on contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/client/TableName.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Objects;
+
+/**
+ * Combination of dbName and tableName fields used
```
```suggestion
 * Combination of database and table names used
```

vdiravka commented on contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java:
```diff
@@ -63,89 +58,38 @@ public Table getTable(String tableName) {
-        tables = Sets.newHashSet(); // empty set.
+        tables = mClient.getTableNamesAndTypes(this.name, schemaConfig.getIgnoreAuthErrors());
+      } catch (TException e) {
+        logger.warn(String.format(
```
Why `String.format`?

ihuzenko replied:
It's invocation of ```warn(String msg, Throwable t)``` which means that stack trace won't be missed in logs. Using string with placeholders ```{}``` and ```warn(String format, Object... arguments)``` most probably will just call ```toString()``` on exception object and stack trace details won't be shown.

vdiravka replied:
I am ok about `e` instead of `e.getCause()`, but in both cases it is the last argument for `warn` method. So not sure what is the difference will a String formatting approach. However WARN level is used by default, users will not have benefit from placeholders. And this is not important. So you can keep your approach here.

ihuzenko commented on contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/client/TableEntryCacheLoader.java:
```diff
@@ -0,0 +1,106 @@
+    }
+    HiveTableWithColumnCache hiveTable = new HiveTableWithColumnCache(table, new ColumnListsCache(table));
+    List<HiveTableWrapper.HivePartitionWrapper> partitionWrappers = partitions.isEmpty()
+        ? null
```
Good catch, the logic was here previously since the class was static nested. So I extracted it and preserved existing logic, but I'll try to use empty list and maybe somewhere else redundant null check will be removed too.

ihuzenko replied:
I've considered possibility to use empty lists and can conclude that doing this will break backward compatibility because ```HiveReadEntry``` is part of JSON serializable  ```HiveScan``` operator, and deserializing empty lists for older drillbit which expect null list may break null dependent checks.

vdiravka replied:
You are right. Not sure that current logic is good, which is based on null checks instead of check list size, but it shouldn't be part of this PR. Thanks for looking into it.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaFilter.java:
```diff
@@ -206,11 +203,11 @@ private Result evaluateHelperFunction(Map<String, String> recordValues, Function
-          }
-          if (exprResult == Result.INCONCLUSIVE) {
-            result = Result.INCONCLUSIVE;
+          switch (exprResult) {
```
Suggested change will break the logic, here is a loop and when invocation of ```evaluateHelper(recordValues, arg)``` returned ```Result.INCONCLUSIVE``` once it's still a chance that next iteration will return false I guess. Previously here was the chunk:
```java
        for(ExprNode arg : exprNode.args) {
          Result exprResult = evaluateHelper(recordValues, arg);
          if (exprResult == Result.FALSE) {
            return exprResult;
          }
          if (exprResult == Result.INCONCLUSIVE) {
            result = Result.INCONCLUSIVE;
          }
        }
```
I see that my change made it more confusing, so I'll revert it back.

vdiravka replied:
oh, sorry, I wonder how I missed it.
Looks like `switch`-`case` approach is more elegant here. Feel free to return you previous version, if you want.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java:
```diff
@@ -920,46 +920,11 @@ public void dropTable(String table) {
-          }
-        }
-      } catch (UnsupportedOperationException e) {
-        logger.debug("The filesystem for this workspace does not support this operation.", e);
```
This deleted code mostly duplicated body of existing ```getViews()``` method. This logging statement also present in the method.

ihuzenko commented:
@vdiravka , I've addressed comments. 

I totally agree with you that refactoring is better to put into separate commits and I'll use this approach in future. 

For show tables authorization improvement was created [DRILL-7151](https://issues.apache.org/jira/browse/DRILL-7151) ticket. 

For caches was changed type of ```tableNamesCache``` to  ```LoadingCache<String, Map<String, Schema.TableType>> ```, previously only names were cached here, also all work with Guava caches was unified under ```HiveMetadataCache``` facade. 

For Drill Hive SASL (Kerberos) connection I didn't introduce changes, related code from ```DrillHiveMetaStoreClientFactory``` was previously in ```DrillHiveMetaStoreClient```.
